platforms/core-runtime/messaging/src/main/java/org/gradle/internal/event/DefaultListenerManager.java

```diff
@@ -179,7 +182,12 @@ private <T> EventBroadcast<T> getBroadcasterInternal(Class<T> listenerClass) {
         synchronized (lock) {
             EventBroadcast<T> broadcaster = Cast.uncheckedCast(broadcasters.get(listenerClass));
             if (broadcaster == null) {
-                broadcaster = new EventBroadcast<T>(listenerClass);
+                if (listenerClass.getAnnotation(StatefulListener.class) != null) {
```

Comment:
❓ Should we be checking for `ParallelListener` here?

Reply:
The listener infrastructure has the concept of a broadcaster and a listener implementation. 

A broadcaster is identified by the interface type that defines the listener interface (i.e. `ProjectComponentObservationListener`). A listener is an _implementation_ of that interface. Listener implementations may implement multiple listener interfaces and therefore may receive events from multiple broadcasters. 

For this reason, there are two sets of locks. The first lock is for a given broadcaster. It controls the registration and de-registration of listeners for a given listener interface. Here, we check for `StatefulListener`, as that specifies behavior over how locks are managed. Specifically, `StatefulListener` here allows us to use an optimized implementation for the broadcaster that allows notifying _listener implementations_ in parallel. This is due to the added restrictions on a listener that a stateful listener requires. 

`ParallelListener` is meant to be added on a _listener implementation_. It defines how an actual implementation of an listener interface is mean to receive events. In this way, we can have one implementation that _is_ capable of receiving events concurrently while simultaneously having an implementation that cannot. 

So to answer your question more briefly, no, since `StatefulListener` controls lock behavior of the broadcaster not the listener implementation